taskcluster/baseclient.py

```diff
@@ -0,0 +1,324 @@
+        log.debug(dict((k, v) for k, v in o.items() if k != 'credentials'))
+
+        if session:
+            self.session = session
+        else:
+            self.session = createSession()
+
+    def makeHawkExt(self):
+        """ Make an 'ext' for Hawk authentication """
+        o = self.options
+        c = o.get('credentials', {})
+        if c.get('clientId') and c.get('accessToken'):
+            ext = {}
+            cert = c.get('certificate')
+            if cert:
+                if six.PY3 and isinstance(cert, six.binary_type):
+                    cert = cert.decode()
+                if isinstance(cert, six.string_types):
+                    cert = json.loads(cert)
+                ext['certificate'] = cert
+
+            if 'authorizedScopes' in o:
+                ext['authorizedScopes'] = o['authorizedScopes']
+
+            # .encode('base64') inserts a newline, which hawk doesn't
+            # like but doesn't strip itself
+            return utils.makeB64UrlSafe(utils.encodeStringForB64Header(utils.dumpJson(ext)).strip())
+        else:
+            return {}
+
+    def _makeTopicExchange(self, exchangeUrl, routingKey, routingKeyPattern):
+        if routingKeyPattern is None:
+            routingKeyPattern = {}
+
+        data = {
+            'exchange': exchangeUrl,
+        }
+
+        # If we are passed in a string, we can short-circuit this function
+        if isinstance(routingKeyPattern, six.string_types):
+            log.debug('Passing through string for topic exchange key')
+            data['routingKeyPattern'] = routingKeyPattern
+            return data
+
+        if not isinstance(routingKeyPattern, dict):
+            errStr = 'routingKeyPattern must eventually be a dict'
+            raise exceptions.TaskclusterTopicExchangeFailure(errStr)
+
+        # There is no canonical meaning for the maxSize and required
+        # reference entry in the JS client, so we don't try to define
+        # them here, even though they sound pretty obvious
+
+        routingKeyParts = []
+        for key in routingKey:
+            if 'constant' in key:
+                value = key['constant']
+            elif key['name'] in routingKeyPattern:
+                log.debug('Found %s in routing key params', key['name'])
+                value = str(routingKeyPattern[key['name']])
+                if not key.get('multipleWords') and '.' in value:
+                    raise exceptions.TaskclusterTopicExchangeFailure(
+                        'Cannot have periods in single word keys')
+            else:
+                value = '#' if key.get('multipleWords') else '*'
+                log.debug('Did not find %s in input params, using %s', key['name'], value)
+
+            routingKeyParts.append(value)
+
+        data['routingKeyPattern'] = '.'.join([utils.toStr(x) for x in routingKeyParts])
+        return data
+
+    def buildSignedUrl(self, requestUrl, expiration=None):
```

Comment:
Ok, thanks for clarifying!  I'll try to add a buildUrl, though I still want to avoid passing around the entire api definition in BaseClient.
The non-exchange objects already have a self.urls with string formatting; I think that, plus allowing for query support, may be sufficient.